cirq-core/cirq/sim/simulator.py

```diff
@@ -656,7 +656,7 @@ def _create_simulator_trial_result(
         self,
         params: study.ParamResolver,
         measurements: Dict[str, np.ndarray],
-        final_simulator_state: TSimulatorState,
+        final_simulator_state: Union[TStepResult, TSimulatorState],
```

Comment:
IIUC, the "step_result OR state" option in `SimulationTrialResult` is required because (1) it needs to support both `SimulatesIntermediateState` and `SimulatesFinalState`, and (2) it's part of the "protected" API.

The same option shouldn't be necessary here: `SimulatesIntermediateState` should always be able to pass a step_result, and this method is not part of the public or protected API. Can we change this to `final_simulator_state: TStepResult`?

Either way, please update the docstring to reflect the new API.

Reply:
The main reason I kept it this way is that unit tests are far easier to write if we allow passing in a final state rather than forcing tests to construct a step result. Is it okay if we keep this and I just update the docstring?

Reply:
The actual code should take priority over the tests. Separately, if the tests are still using `TSimulatorState` that means they're not exercising the behavior used in the actual code - the tests should be updated.

(On a tangential note: in addition to verifying the code, tests also provide an example of how the code should be used. If the tests are based on deprecated behavior, they no longer fulfill this purpose.)